### DISCUSSION TITLE
is there a good templating engine for Pocketbase JSVM + htmX ?

### DISCUSSION THREAD

**Discussion by gokulgseven** [pocketbase/pocketbase#4285] [Q&A] [answered]
A really powerful combination. 

but having some trouble getting things set up correctly. 
only problem I am currently facing is good syntax highlighting for html code inside JSMV (*pb.js file)
can someone suggest some vsCode extensions for this ?
and a good templating library out there that you have tried that works well with pocketbase JSVM 

will https://developer.hashicorp.com/nomad/tutorials/templates/go-template-syntax work with JSMV ?

**Comment by gokulgseven** [comment]
found a syntax highlighting extention that works : leet-html

now just looking for a templating system that works inline inside pb.js 

this is for go and looks amazing: 
https://github.com/a-h/templ 

without the template engine i am having to do something like this to create a table
```
let a = `
            <table border='1'>
                <tr>
                    <th>Name</th>
                    <th>Age</th>
                </tr>
                ${people.map(person => 
                    `<tr>
                        <td>${person.name}</td>
                        <td>${person.age}</td>
                    </tr>`).join("")}
            </table>`
```
and in this syntax highlighting for html doesn't work , inside the arr.map()

**Reply by gokulgseven** [reply]
htmx does have a https://htmx.org/extensions/client-side-templates/ but they dont make sense , it should be all server side {thats the whole point of htmx}

**Comment by pchasco** [comment]
As PocketBase is just a Go application, wouldn't you just use any Go templating engine? Go has a standard template engine, text/template, which is probably fine for anyone's needs. There are some others that may offer better performance and a different API/syntax which you may prefer over the others.

Client side templates make sense for HTMX when you are pulling data from services that are outside your control, or that do not make sense to render server side -- for example, if you have a service that already returns JSON. You could marshal that data through your web server and transform the JSON to HTML server-side, but in some cases the drawbacks may outweigh the benefits (additional latency, additional development time, etc.).

**Reply by gokulgseven** [reply]
yes , i was looking for a templating engine for javascript that is as good as the go one cause i am not yet familiar with go. maybe its time i start learing go .

**Reply by pnmcosta** [reply]
@gokulgseven this https://github.com/pocketbase/pocketbase/blob/master/tools/template/registry.go maybe a great resource to get started.

I rolled my own collection driven templates solution and I use the registry to render my templates. It's been awesome to work with.

**Reply by gokulgseven** [reply]
its in .go, why is go so popular now , every where i see there are people using .go . 
going to have to learn it! 

Anyway for the curently thing i will just make something myself in javascript . all i need is loops and conditions will upload the code here when its made, for the next person.

**Reply by ganigeorgiev** [reply, marked as answer]
@gokulgseven You can use the Go templates with the JSVM - https://pocketbase.io/docs/js-rendering-templates/.

**Reply by gokulgseven** [reply]
missed that in the documentation, thank you.

**Comment by pchasco** [comment]
Oh, yes, ok. If your plan is to use HTMX, you should do your HTML
generation on the server except when there are exceptional cases.

That being said, I have used Handlebars and Mustache on the client side and
they are both easy to use.

**Comment by benallfree** [comment]
Did you see #5444 

https://pocketpages.dev

**Reply by benallfree** [reply]
Are you referring to https://htmx.org/? I haven't looked at it much, but it seems like it's a purely client-side technology, right?

**Reply by davidbarton** [reply]
I have briefly read through https://pocketpages.dev/docs/overview and it seems to me that HTMX with https://pocketbase.io/docs/js-rendering-templates/ is a better option in most cases.

@benallfree If you did not experiment with HTMX yet I would suggest you give it a try since you may like it or at minimum take some ideas from it.

**Reply by benallfree** [reply]
@davidbarton I've been looking at it along with AlpineJS. For me the built-in PocketBase templating was too verbose and raw. PocketPages adds a file-based router, processes templates in Markdown and EJS which is hugely more convenient than Go's template syntax (especially for someone who isn't doing anything else in Go). Options are a good thing :)

**Reply by davidbarton** [reply]
@benallfree Yeah and I think they can be complimentary. The router is definitely a plus.

**Comment by davidbarton** [comment]
I think PocketBase & HTMX is a match made in heaven. The biggest drawback of HTMX usage in PocketBase now is that reusing the existing API endpoints is not straightforward. You may think why would I want to reuse them? The answer is simple. The existing API already does solve the whole data (and auth!) problem and it would be pointless to re-implement it yet again only to return HTMX instead of JSON with exactly same data (and same auth).

Basically there are two options to (ab)use the existing API structure to return HTMX fragments instead of JSON.
1. You can define a proxy for each endpoint. It forwards the request to the API endpoint and uses the response as variables to render the template fragment.
2. You can define a custom middleware that would capture the response body being written and allow to mutate it before the response is send. In our case, use the json as variables to render template fragment and use it as new response.

With better support for any of these, using PocketBase with HTMX (and [json plugin](https://v2-0v2-0.htmx.org/extensions/json-enc/)) is as easy as defining new collections and writing html fragments. No other code needed.

Here is some discussion about rewriting the response in echo https://github.com/labstack/echo/discussions/1992 & https://github.com/labstack/echo/discussions/2375 & https://github.com/labstack/echo/issues/658

**Reply by pnmcosta** [reply]
Good observations. However if you use PB as a framework https://pocketbase.io/docs/use-as-framework/ you can leverage 99% of the internal PB APIs that would mirror the collection routes. 

I have certainly done so in my project and have been serving HTMX in production - although not authed, as don't have the need for it on my use cases.

**Reply by benallfree** [reply]
Do you mean return HTML from an API route? If so, how would you vary the formatting based on how it's needs to be displayed?

**Reply by pnmcosta** [reply]
Here's a very small (and truncated) example/snippet of a custom handler where I render `html/template` views that have been embedded:

```
//go:embed views/*
var viewsFS embed.FS

type handler struct {
	app      core.App
	registry *template.Registry
}

// bind where applicable on your app load
func BindHandler(app core.App) {
	registry := template.NewRegistry()

	app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
		h := handler{
			app:      app,
			registry: registry,
		}

		routes := e.Group(
			"/my-routes",
			apis.ActivityLogger(app),
			// any other middlewares from pb or custom
		)

		// just a example route
		routes.GET("", h.get)

		return nil
	})
}

func (h handler) get(c echo.Context) error {
	views := []string{"views/layout.tmpl", "views/my-view.tmpl"}
	renderData := struct {
		LayoutField string
		MyViewField string
	}{
		LayoutField: "a-value",
		MyViewField: "b-value",
	}
	return h.render(c, views, renderData)
}

func (h handler) render(c echo.Context, views []string, data any) error {
	html, err := h.registry.LoadFS(viewsFS, views...).Render(data)
	if err != nil {
		return apis.NewNotFoundError("", err)
	}

	return c.HTML(http.StatusOK, html)
}

```

**Reply by pnmcosta** [reply]
@benallfree could elaborate on

> you vary the formatting based on how it's needs to be displayed?

 with an example? not sure I follow 😅

**Reply by pnmcosta** [reply]
Oh I think I know what you mean, so you want the same collection routes to return both JSON or HTML? I don't think that's possible, and outside the scope of PB (AFAIK).

But you can certainly mirror them 😅 if required. My example above is for a custom route of course.

**Reply by davidbarton** [reply]
@benallfree Yes I mean to return HTML from an API route.

Do you talk about different viewports or engines or what do you mean? For everything that is not a web browser I would still use the API as is (the middleware is used only with HTMX header present). For different browser viewports something like Tailwind responsive styles would do the job.

**Reply by pnmcosta** [reply]
In my specific case and needs, PB was the perfect backend, my product is a Shopify App, so we needed to use React on the Admin, but on the extension/plugins we setup on storefronts we're using HTMX cause of it's lightweight nature.

**Reply by davidbarton** [reply]
@pnmcosta That is certainly an option but the huge benefit is that I do not need to write any of that with such setup. All the work has already been done including the auth and can be used as is. With PocketBase middleware to handle the body rewrite, the minimal usage can be as simple as this:

```
reweriteBody([{
  apiMethod: 'GET',
  apiRoute: '/api/users',
  variableMapping: (x) => {...x},
  templates: ['layout.html', 'userList.html'],
}])
```

**Reply by pnmcosta** [reply]
That is very interesting, where can I find out more about that `rewriteBody` middleware?

**Reply by davidbarton** [reply]
That is just an example. It would essentially do one of these two rewrite options I described in https://github.com/pocketbase/pocketbase/discussions/4285#discussioncomment-10534293

**Reply by pnmcosta** [reply]
Ah ok, got it, thank you.

If you're actively working on it, please consider Gani's refactor, more info on https://github.com/pocketbase/pocketbase/discussions/4355

Specifically that Echo should be dropped in favor of the std library. And I'm also not sure if there would be a method to inject custom middlewares to the collection routes, but I may be mispeaking.

**Reply by davidbarton** [reply]
Did not know about Echo going away. Thanks for the link. It means the proxy option should be used.
